[Shaine Goh Si Hui] iP

soloplxya opened the pull request:
# Duke 
Here's a famous quote to start of this thread:
> “Your mind is for having ideas, not holding them.” – David Allen 

## So what exactly is Duke and how can Duke help you?
Duke frees your mind of having to remember things you need to do. It's,
-  Text-based
-  Easy to learn
-  ~~FAST~~ Incredibly FAST to use!!

All you need to do is,
1. check out this [link](https://github.com/soloplxya/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you

Needless to say, 
IT IS **FREE** for USE!! :shipit:

These are the features that we have developed so far: 
- [x] Add tasks
- [x] Finding tasks
- [ ] Managing deadlines 
- [ ] An actual GUI for the tool is also coming up soon!! 


Feel free to edit the source code if you wish to as well! This is what the main method looks like : 
```  java
public static void main(String[] args) throws DukeException {
        new Duke("./data", "./data/duke.txt").run();
}
```

Commit:
Add Gradle support

soloplxya committed:
Level 1: Greet, Echo and Exit

soloplxya committed:
Level 2: Add,List

soloplxya committed:
Level 3: Mark As Done (With Extension)

soloplxya committed:
Level 4: Todos, Events and Deadlines (With A-Inheritance)

soloplxya committed:
A-TextUiTesting: Baseline testcases

soloplxya committed:
Level 5: Handle Errors (With A-Exceptions)

soloplxya committed:
Level 6: Delete (with A-Collections)

soloplxya committed:
Chore: Added more test inputs

soloplxya committed:
A-Enums: Added Enums for CommandTypes

soloplxya committed:
Add ability for Duke to save tasks locally (Level 7)

soloplxya committed:
Convert event and deadline dates to Date/DateTime format (Level 8)

soloplxya committed:
Merge branch 'branch-level-8'

soloplxya committed:
Resolve conflicts from merging level 7 and level 8 branches

soloplxya committed:
Create TaskList, TextUi, Storage, Parser classes and packaging them into duke/task packages

soloplxya committed:
Add JUnit testcases for Ui, Event and Todo classes (A-Junit)

soloplxya committed:
Add javadocs for all class methods and constructors (A-JavaDoc)

soloplxya committed:
Edit code to comply with Code Standard (A-CodingStandard)

soloplxya committed:
Implement find functionality (Level-9)

soloplxya committed:
Merge branch 'branch-A-JavaDoc'

soloplxya committed:
Resolve merge conflict for merging branch-JavaDocs and branch-CodeStyle

soloplxya committed:
Resolve merge conflict for merging branch-Level9

soloplxya committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

soloplxya committed:
Set up Gradle (A-Gradle)

soloplxya committed:
Merge branch 'branch-A-Gradle'

soloplxya committed:
Add CheckStyle configs and update files to fit code style (A-CheckStyle)

soloplxya committed:
Merge branch 'branch-A-CheckStyle'

soloplxya committed:
Add GUI to Duke using JavaFX (Level-10)

soloplxya committed:
Merge branch 'branch-Level-10'

cheekean5848 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,22 @@
+     * @return A string of the error message
+     */
+    @Override
+    public String toString() {
+        return this.getMessage();
+    }
```
Is overriding getMessage better?

cheekean5848 commented on src/main/java/duke/TextUi.java:
```diff
@@ -0,0 +1,46 @@
+
+public class TextUi {
+    private static final String DIVIDER = "===================================================";
+    private static final String logo = " ____        _\n"
```
isn't the logo variable should be capitalise since it is constant

cheekean5848 reviewed (commented):
Overall is good, I like how you name your action as verbs and personalise Duke to japanese style.
Keep it going

soloplxya committed:
Add Commands package and javadocs

soloplxya committed:
Add assertions (A-Assertions)

glennljw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,131 @@
+        }
+    }
+
+    /**
+     * Method that takes in a command Type and index and performs the appropriate task action
+     * @param commandType type of command user has inputted (i.e. bye, list...)
+     * @param index taskId
+     * @throws DukeException if task cannot be found within the tasklist
+     * @throws NumberFormatException if taskId is not a number.
+     */
+    public Command taskAction(CommandType commandType, String index) throws DukeException, NumberFormatException {
```
Should there be a @return for the JavaDocs here?

glennljw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,131 @@
+            }
+
+            switch (commandType) {
+            case FIND:
+                return new Find(inputArray[1]);
+            case DELETE:
+                return taskAction(CommandType.DELETE, inputArray[1]);
+            case UNMARK:
+                return taskAction(CommandType.UNMARK, inputArray[1]);
+            case MARK:
+                return taskAction(CommandType.MARK, inputArray[1]);
```
Perhaps you can consider making inputArray[1] a local variable in the method so as to give it a descriptive name for easy reference?

glennljw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,131 @@
+package duke;
+import duke.commands.AddDeadline;
```
Should there be a line break after the package statement before the import statements?

glennljw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,44 @@
+package duke;
+import tasks.TaskList;
```
Should there be a line break after the package statement before the import statements?

glennljw reviewed (commented):
Overall great job! I like your implementation! 😄

soloplxya committed:
Refactor code to improve readability (A-CodeQuality)

- Currently, the Parser class has long methods and numerous error messages which makes the code hard to understand. Additionally, the taskList class also has message strings in various return statements which should be handled in the TextUi class. Furthermore, all exceptions are currently classified as DukeExceptions which makes it difficult to tell apart the various exception types that needs to be handled.

- It is necessary to refactor the current code so as to make the code more readable and easier to refactor in the future.

- Long methods in the Parser class have been refactored using SLAP. Additionally, instead of returning the ui messages directly in the TaskList class, these ui messages are now handled by methods in the TextUi class. Lastly, instead of using DukeException to handle all exceptions, the exceptions have been broken down into specific exception classes.

- Inheritance is preferable for the exceptions so as to attain polymorphic behaviour when rendering the error messages. It is also appropriate to use inheritance in this scenario since "InvalidCommandException" and "TaskException" are DukeExceptions.

soloplxya committed:
Commit remaining staged files to branch(A-CodeQuality)

- Currently, the Parser class has long methods and numerous error messages which makes the code hard to understand. Additionally, the taskList class also has message strings in various return statements which should be handled in the TextUi class. Furthermore, all exceptions are currently classified as DukeExceptions which makes it difficult to tell apart the various exception types that needs to be handled.

- It is necessary to refactor the current code so as to make the code more readable and easier to refactor in the future.

- Long methods in the Parser class have been refactored using SLAP. Additionally, instead of returning the ui messages directly in the TaskList class, these ui messages are now handled by methods in the TextUi class. Lastly, instead of using DukeException to handle all exceptions, the exceptions have been broken down into specific exception classes.

- Inheritance is preferable for the exceptions so as to attain polymorphic behaviour when rendering the error messages. It is also appropriate to use inheritance in this scenario since "InvalidCommandException" and "TaskException" are DukeExceptions.

soloplxya committed:
Merge pull request #1 from soloplxya/branch-A-Assertions

Add assertions (A-Assertions)

Merge A-Assertions branch to master

soloplxya committed:
Merge branch 'master' of https://github.com/soloplxya/ip

soloplxya committed:
Merge branch 'master' into branch-A-CodeQuality

soloplxya committed:
Merge pull request #2 from soloplxya/branch-A-CodeQuality

Refactor code to improve readability (A-CodeQuality)

Merge branch-A-CodeQuality to master

soloplxya committed:
Merge branch 'master' of https://github.com/soloplxya/ip

soloplxya committed:
Resolve merge conflicts

soloplxya committed:
Set up continuous integration (A-CI)

soloplxya committed:
Update gradle.yml file to exclude IO-redirection checks (A-CI)

Duke does not automate building and testing after each code change.

It is important for Duke to be able to do so as it allows the developer to identify bugs faster. It also makes the process of testing more seamless and efficient.

To facilitate automated testing and building, let's add automated workflows to test every push and pull request to the repository.

Setting up CI will allow easier shipment and deployment of code.

soloplxya committed:
Fix checkstyle errors

soloplxya committed:
Add undo command for Duke

An undo command does not exist in the current duke.

Such a command is needed as users may want to undo an action that they made by mistake.

Add an undo command to allow users to undo add, mark, unmark and delete commands.

As all commands have different ways of undoing, it is better to override an undo method in the Command class and make use of polymorphism to customize each undo function.

soloplxya committed:
Merge pull request #3 from soloplxya/branch-BCD-Extension

Add undo command for Duke

soloplxya committed:
Improve Duke's GUI to suit a Valentine's Day theme (A-BetterGui)

soloplxya committed:
Add screenshot of Ui to docs folder

soloplxya committed:
Update README.md

soloplxya committed:
Update typos in README.md

soloplxya committed:
Update filename for Ui.png and header comments in TaskList.java

soloplxya committed:
Replace Ui.png file

soloplxya committed:
Rename Ui.png file

soloplxya committed:
Delete old Ui.png file

soloplxya committed:
Clean up javadocs

soloplxya committed:
Make changes to README.md

soloplxya committed:
Fix long methods and code standard violations

soloplxya committed:
Add jUnit tests for deadlines

soloplxya committed:
Set theme jekyll-theme-midnight

soloplxya committed:
Set theme jekyll-theme-hacker

soloplxya committed:
Add credit for reused code and update find feature in README.md

soloplxya committed:
Merge branch 'master' of https://github.com/soloplxya/ip

soloplxya committed:
Set theme to default

soloplxya committed:
Update readme and javadocs

soloplxya committed:
Update Ui.png

soloplxya committed:
Update README.md

soloplxya committed:
Update README.md

soloplxya committed:
Clean up file path code

soloplxya committed:
Merge branch 'master' of https://github.com/soloplxya/ip

soloplxya committed:
Update README.md
